src/main/scala/io/iohk/ethereum/merklepatriciatrie/MerklePatriciaTrie.scala

```diff
@@ -0,0 +1,486 @@
+            val temporalExtensionNode = ExtensionNode(sharedKeySuffix, next, hashFn)
+            val NodeInsertResult(newBranchNode: BranchNode, toDeleteFromStorage, toUpdateInStorage) = put(temporalExtensionNode, searchKey.drop(ml), value)
+            val newExtNode = ExtensionNode(sharedKeyPrefix, newBranchNode, hashFn)
+            NodeInsertResult(
+              newNode = newExtNode,
+              toDeleteFromStorage = node +: toDeleteFromStorage,
+              toUpdateInStorage = newExtNode +: toUpdateInStorage
+            )
+        }
+      case branchNode@BranchNode(children, optStoredValue, _) =>
+        if (searchKey.isEmpty) {
+          // The key is empty, the branch node should now be a terminator node with the new value asociated with it
+          val newBranchNode = BranchNode(children, Some(value), hashFn)
+          NodeInsertResult(
+            newNode = newBranchNode,
+            toDeleteFromStorage = Seq(node),
+            toUpdateInStorage = Seq(newBranchNode)
+          )
+        }
+        else {
+          // Non empty key, we need to insert the value in the correct branch node's child
+          val searchKeyHead: Int = searchKey(0)
+          val searchKeyRemaining = searchKey.tail
+          if (children(searchKeyHead).isDefined) {
+            // The associated child is not empty, we recursively insert in that child
+            val NodeInsertResult(changedChild, toDeleteFromStorage, toUpdateInStorage) =
+            put(getChild(branchNode, searchKeyHead, dataSource).get, searchKeyRemaining, value)
+            val newBranchNode = branchNode.updateChild(searchKeyHead, changedChild, hashFn)
+            NodeInsertResult(
+              newNode = newBranchNode,
+              toDeleteFromStorage = node +: toDeleteFromStorage,
+              toUpdateInStorage = newBranchNode +: toUpdateInStorage
+            )
+          }
+          else {
+            // The associated child is empty, we just replace it with a leaf
+            val newLeafNode = LeafNode(searchKeyRemaining, value, hashFn)
+            val newBranchNode = branchNode.updateChild(searchKeyHead, newLeafNode, hashFn)
+            NodeInsertResult(
+              newNode = newBranchNode,
+              toDeleteFromStorage = Seq(node),
+              toUpdateInStorage = Seq(newLeafNode, newBranchNode)
+            )
+          }
+        }
+    }
+  }
+
+  private def remove(node: Node, searchKey: Array[Byte]): NodeRemoveResult = (node, searchKey.isEmpty) match {
+    // They key matches a branch node but it's value doesn't match the key
+    case (BranchNode(children, None, _), true) => NodeRemoveResult(hasChanged = false, maybeNewChild = None)
+    // We want to delete Branch node value
+    case (BranchNode(children, _, _), true) =>
+      // We need to remove old node and fix it because we removed the value
+      val fixedNode = fix(BranchNode(children, None, hashFn), dataSource, Seq(), hashFn)
```

Comment:
@adamsmo because one is used within BranchNode's apply and the second one is used by fix fn when creating a `LeafNode`. An alternative is to make Node's hashFn public and take it from there, what do you think?